### PR TITLE
Use testcontainers networking capabilities when using @QuarkusIntegrationTest + DevServices

### LINKED ISSUE #18914: QuarkusIntegrationTest: docker option --net=host not supported for windows
In Quarkus 2.1.0 the docker run command has changed to bind to the host network instead. This property, however, is invalid on windows (and mac?). Nor does this work properly if the container is started from another container (for example on our gitlab runner).

https://github.com/quarkusio/quarkus/blob/d0555c781c15cbc684684b3d79f98f2507658a94/test-framework/common/src/main/java/io/quarkus/test/common/DefaultDockerContainerLauncher.java#L81

Obviously the old behavior had major issues as well, so I thought of a solution that will work in all cases:

Why not run the docker image (containing the application) using testcontainers as well? 
- All the test containers should be started on the same network (`withNetwork(Network.SHARED)`, see the Kafka dev service)
- The dev services set a Network Alias (`withNetworkAliases("name-" + Base58.randomString(5))`)
- The `DevServicesNativeConfigResultBuildItem` should contain both 'regular' as 'container-network' values, for example:
  - `kafka.bootstrap.servers` has `PLAINTEXT://localhost:49215` and `PLAINTEXT://kafka-x42bz:9092,`
  - `quarkus.mongodb.connection-string` has `mongodb://localhost:49213/test` and `mongodb://mongo-8ux2l:27017/test`

When running the testcontainer containing the application, we would provide the second set of values while all others receive the first set.

### PR DESCRIPTION
**Pull request by geoand**
This is done in order to get us off of host networking (which only works on Linux).
The way this works is that when `@QuarkusIntegrationTest` is meant to launch a container
and DevServices are needed, a build item is pushed into the build that forces
the DevServices to launch in TestContainers' `SHARED` Network.
The id of network is then used when launching the application as a container.

This change only makes a difference for the container launch case of `@QuarkusIntegrationTest`.

Fixes: #18914

### CODE REVIEW

**Comment by geoand** [comment]
@stuartwdouglas mind taking a look to validate the approach before I move to update the rest of the DevServices?

**Comment by stuartwdouglas** [comment]
I don't really like the complexity it adds to all the DevServices, although I also can't really think of any better options. 

Is there no way to have this logic in a common place somehow? It looks like it might be specific to each container, which means we really should test every DevServices in both shared and non-shared mode.

**Comment by geoand** [comment]
> I don't really like the complexity it adds to all the DevServices, although I also can't really think of any better options.
> 

I agree and it's why I put this off for so long...

> Is there no way to have this logic in a common place somehow? It looks like it might be specific to each container, which means we really should test every DevServices in both shared and non-shared mode.

I don't think so as we need to manipulate the returned URL of each container. 
We could have the hostname generation be common, but that wouldn't really reduce the complexity much.

As for testing, we can convert some of the integration tests into container builds and use them.

**Comment by geoand** [comment]
The PR should be ready to go now

**Comment by sberyozkin** [comment]
Hi @geoand Can you please update `KeycloakDevServicesProcessor` as well ?

**Comment by sberyozkin** [comment]
Sorry, you already did :-)

**Comment by geoand** [comment]
> I updated all the DevServices and added a bunch of tests

I should also mention that these tests exercise both the shared network and default code paths because dev-services comes into play for them both for surefire and failsafe

**Comment by geoand** [comment]
@stuartwdouglas can I get a review of this? Since it touches all the dev services, there is a good chance it could get conflicts soon

**Comment by bdevreugd-vialis** [comment]
I do have one comment, because all the communication is now network internal, the issue of: https://github.com/quarkusio/quarkus/issues/18854 kind of returns.

I would still like to be able to access the dev services from the test code itself, maybe also expose a property with postfix `.from-host` that will contain the original URL (with the mapped port)?

**Comment by geoand** [comment]
> I do have one comment, because all the communication is now network internal, the issue of: #18854 kind of returns.
> 
> I would still like to be able to access the dev services from the test code itself, maybe also expose a property with postfix `.from-host` that will contain the original URL (with the mapped port)?

Good point. Let's leave that for a next PR, so please open an issue once this goes in

**Comment by geoand** [comment]
Ideally we would backport this one, but given the scope of the change I think it's safer not to, so I am removing the label

**Review by stuartwdouglas — APPROVED** [review]

File: extensions/devservices/db2/src/main/java/io/quarkus/devservices/db2/deployment/DB2DevServicesProcessor.java [hunk 1]
```diff
@@ -57,4 +53,53 @@ public void close() throws IOException {
...
+                    .asCompatibleSubstituteFor(DockerImageName.parse("ibmcom/db2")));
+            this.fixedExposedPort = fixedExposedPort;
+            this.useSharedNetwork = useSharedNetwork;
+        }
+
+        @Override
+        protected void configure() {
+            super.configure();
+
+            if (useSharedNetwork) {
```
Comment (stuartwdouglas) [inline extensions/devservices/db2/src/main/java/io/quarkus/devservices/db2/deployment/DB2DevServicesProcessor.java:73]: Is there any chance this code can be moved to a shared utility method so it is not copied everywhere?

Comment (geoand) [reply extensions/devservices/db2/src/main/java/io/quarkus/devservices/db2/deployment/DB2DevServicesProcessor.java:73]: I'll look into it

Comment (geoand) [reply extensions/devservices/db2/src/main/java/io/quarkus/devservices/db2/deployment/DB2DevServicesProcessor.java:74]: I moved the shared network stuff into a utility method. The rest isn't that straighforward as some of the methods are protected

**Pull request merged**